src/types.rs

```diff
@@ -475,3 +506,157 @@ impl<'buf> RecvMsgOut<'buf> {
         self.header.flags
     }
 }
+
+/// [CancelBuilder] constructs match criteria for request cancellation.
+///
+/// The default behavior of the [CancelBuilder] is cancel all in-flight requests,
+/// however this can be refined to cancel only requests which reference a file descriptor
+/// or have a specific `user_data` value.
+///
+/// Additionally, [CancelBuilder::all] can be set to match all requests matching the given
+/// criteria, rather than just the first one that would have matched.
+///
+/// ### Examples
+///
+/// ```
+/// use io_uring::types::{CancelBuilder, Fd, Fixed};
+///
+/// // Match all in-flight requests.
+/// CancelBuilder::new();
+///
+/// // Match a single request with user_data = 42.
+/// CancelBuilder::new().user_data(42);
+///
+/// // Match a single request with fd = 42.
+/// CancelBuilder::new().fd(Fd(42));
+///
+/// // Match a single request with fixed fd = 42.
+/// CancelBuilder::new().fd(Fixed(42));
+///
+/// // Match all in-flight requests with user_data = 42.
+/// CancelBuilder::new().user_data(42).all();
+/// ```
+#[derive(Debug)]
+pub struct CancelBuilder {
+    pub(crate) flags: AsyncCancelFlags,
+    pub(crate) user_data: Option<u64>,
+    pub(crate) fd: Option<sealed::Target>,
+}
+
+impl CancelBuilder {
+    /// Create a new [CancelBuilder].
+    ///
+    /// The default behavior of the [CancelBuilder] is cancel all in-flight requests.
+    /// This behavior can be further refined by calling the builder methods on the [CancelBuilder].
+    pub const fn new() -> Self {
+        Self {
+            flags: AsyncCancelFlags::ANY,
+            user_data: None,
+            fd: None,
+        }
+    }
+
+    /// Refine the match criteria to match only requests which contain the
+    /// provided `user_data`.
+    ///
+    /// This is mutually exclusive with [CancelBuilder::fd](#method.fd).
+    pub fn user_data(mut self, user_data: u64) -> Self {
+        // Unset the ANY flag because we want to refine the match criteria to a subset of requests.
+        self.flags.set(AsyncCancelFlags::ANY, false);
```

Comment:
I'm a bit curious which is better to use insert/remove vs set .

Reply:
I'm not sure, tbh I didn't realize insert/remove existed but it seems a bit clearer to me, so I switched to using that instead of set.